### ISSUE TITLE
Add @since to the coding standard

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2500]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2500|
|Type|enhancement|
|Created|2007-03-01 17:59:32Z|
|Branch|https://github.com/twisted/twisted/tree/version-coding-std-2500|

The API documentation should be marked up with version information.  Using an epytext field for this probably makes the most sense, but I don't really care so much about the actual mechanism, as long as the information is present.

One annoying factor is that we don't always know the next version number for the full development time leading up to it.  It might be nice to be able to drop a marker value in and have part of the release process replace it with a real value.

There should be a similar standard for deprecations.

<details><summary>Searchable metadata</summary>

```
trac-id__2500 2500
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_version_coding_std_2500 branches/version-coding-std-2500
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1172771972000000 1172771972000000
changetime__1220912968000000 1220912968000000
version__None None
owner__ 
cc__thijs cc__radix cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

I suggest using @deprecate. I could add the @version information but I'd need to have a list of what was added when, or perhaps mark everything newer than > copyright 2004 as v2.5, and stuff that is older as v2.4 or something?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [24695]) Branching to 'version-coding-std-2500'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

Hmm.

The idea here is definitely still a good one.  The preferred implementation might be different now than when the ticket was filed, though.  [#3266](https://github.com/twisted/twisted/issues/3266) describes an API for deprecating things; this would provide all the information necessary for the API docs.  I wonder if we should do something similar for `@since`.  This would basically mean decorating every API in Twisted.  That's bad, I guess.  So maybe we should stick with `@since` but drop `@deprecated` in favor of [#3266](https://github.com/twisted/twisted/issues/3266).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

I don't think I intended to assign this to me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_6):
>  So maybe we should stick with `@since` but drop `@deprecated` in favor of [#3266](https://github.com/twisted/twisted/issues/3266).

I removed the part about @deprecated in r24730, putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

updating title

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

The docs say _@since 8.1_ but I guess this should be _@since: 8.1_.  Otherwise looks good.  Please fix and merge.  Note that this information won't be rendered in a very nice way until pydoctor is [slightly improved](https://bugs.launchpad.net/pydoctor/+bug/267943).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [24749]) Merge version-coding-std-2500: Add Versioning section, describing the '@since' tag, to the coding standard document.

Author: thijs
Reviewer: exarkun
Fixes: [#2500](https://github.com/twisted/twisted/issues/2500)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
